File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/TextChangeAnimator.java
package com.reactnativenavigation.views.element.animators;

import android.animation.Animator;
import android.graphics.Point;
import android.graphics.Rect;
import android.view.View;
import android.widget.TextView;

import com.facebook.react.views.text.ReactTextView;
import com.reactnativenavigation.utils.ViewUtils;
import com.reactnativenavigation.views.ComponentLayout;
import com.shazam.android.widget.text.reflow.ReflowTextAnimatorHelper;

import androidx.annotation.NonNull;

import static com.reactnativenavigation.utils.TextViewUtils.getTextSize;

public class TextChangeAnimator extends PropertyAnimatorCreator<ReactTextView> {

    public TextChangeAnimator(View from, View to) {
        super(from, to);
    }

    @Override
    protected boolean shouldAnimateProperty(ReactTextView fromChild, ReactTextView toChild) {
        Point fromXy = ViewUtils.getLocationOnScreen(from);
        Point toXy = ViewUtils.getLocationOnScreen(to);
        return getTextSize(fromChild) != getTextSize(toChild) ||
               !fromXy.equals(toXy.x, toXy.y);
    }

    @Override
    public Animator create() {
        return new ReflowTextAnimatorHelper
                .Builder((TextView) from, (TextView) to)
                .calculateDuration(false)
                .setBoundsCalculator(view -> {
                    int[] loc = new int[2];
                    view.getLocationInWindow(loc);
                    @NonNull View parent = ViewUtils.findParent(view, ComponentLayout.class);
                    return new Rect(loc[0], loc[1] - parent.getTop(), loc[0] + view.getWidth(), loc[1] + view.getHeight());
                })
                .buildAnimator();
    }
}

File: lib/android/app/src/main/java/com/reactnativenavigation/views/element/animators/TextChangeAnimator.kt
package com.reactnativenavigation.views.element.animators

import android.animation.Animator
import android.graphics.Rect
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.view.updateMargins
import com.facebook.react.views.text.ReactTextView
import com.reactnativenavigation.parse.SharedElementTransitionOptions
import com.reactnativenavigation.utils.TextViewUtils
import com.reactnativenavigation.utils.ViewUtils
import com.reactnativenavigation.views.element.SharedElementTransition
import com.shazam.android.widget.text.reflow.ReflowTextAnimatorHelper

class TextChangeAnimator(private val transition: SharedElementTransition, from: View, to: View) : PropertyAnimatorCreator<ReactTextView>(from, to) {
    override fun shouldAnimateProperty(fromChild: ReactTextView, toChild: ReactTextView): Boolean {
        val fromXy = ViewUtils.getLocationOnScreen(from)
        val toXy = ViewUtils.getLocationOnScreen(to)
        return TextViewUtils.getTextSize(fromChild) != TextViewUtils.getTextSize(toChild) ||
                !fromXy.equals(toXy.x, toXy.y)
    }

    override fun create(options: SharedElementTransitionOptions): Animator {
        (to.layoutParams as ViewGroup.MarginLayoutParams).apply {
            updateMargins(top = to.top + transition.topInset)
        }
        return ReflowTextAnimatorHelper.Builder(from as TextView, to as TextView)
                .calculateDuration(false)
                .setBoundsCalculator { view: View ->
                    val loc = IntArray(2)
                    view.getLocationInWindow(loc)
                    val topInset = if (view == to) this.transition.topInset else 0
                    Rect(
                            loc[0],
                            loc[1] + topInset,
                            loc[0] + view.width,
                            loc[1] + view.height
                    )
                }
                .setTextColorGetter {
                    TextViewUtils.getTextColor(it)
                }
                .buildAnimator()
                .setDuration(options.getDuration())
    }
}